File: app/src/main/java/com/testosterolapp/unrd/messages/MessagesScreenAdapter.java
package com.testosterolapp.unrd.messages;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;
import com.testosterolapp.unrd.R;
import com.testosterolapp.unrd.data.Data;
import com.testosterolapp.unrd.db.DaoRepository;
import com.testosterolapp.unrd.util.QuickContactDivot;

import java.util.ArrayList;

public class MessagesScreenAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private ArrayList<Data> messagesList;

    private static int TYPE_SENT = 1;
    private static long MAIN_CHAR_ID = 218;

    private Long mainCharId;

    private Context mContext;
    static private Drawable sDefaultContactImage;
    private DaoRepository daoRepository;

    public MessagesScreenAdapter(Context context, ArrayList<Data> dataList, Long mainCharId) {
        this.mContext = context;
        this.messagesList = dataList;
        this.mainCharId = mainCharId;
        this.daoRepository = new DaoRepository(context);
        if (sDefaultContactImage == null) {
            sDefaultContactImage = context.getResources().getDrawable(R.drawable.ic_contact_picture);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (messagesList != null) {
            if (position <= messagesList.size()) {
                Data message = messagesList.get(position);
                if (message != null) {
                    if (message.getCharacter_id() != null && message.getCharacter_id() != (MAIN_CHAR_ID)) {
                        return 2;
                    } else {
                        return TYPE_SENT;
                    }
                }
                return TYPE_SENT;
            }
        }
        return TYPE_SENT;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (i != TYPE_SENT) {
            //SENT UI
            View itemView = LayoutInflater.from(viewGroup.getContext())
                    .inflate(R.layout.message_list_sent_item, viewGroup, false);
            return new SentViewHolder(itemView);
        } else {
            //RECEIVE UI
            View itemView = LayoutInflater.from(viewGroup.getContext())
                    .inflate(R.layout.message_list_receive_item, viewGroup, false);
            return new ReceiveViewHolder(itemView);
        }
    }

    @Nullable
    protected Data getItem(int position) {
        return messagesList.get(position);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (position <= -1) {
            return;
        }

        Data data;
        try {
            data = getItem(position);
        } catch (IndexOutOfBoundsException e) {
            return;
        }

        if (holder instanceof SentViewHolder) {
            if (data != null) {
                ((SentViewHolder) holder).setupViewHolder(data);
            }
        } else if (holder instanceof ReceiveViewHolder) {
            if (data != null) {
                ((ReceiveViewHolder) holder).setupViewHolder(data);
            }
        }

    }


    @Override
    public int getItemCount() {
        return messagesList.size();
    }

    class SentViewHolder extends RecyclerView.ViewHolder {

        TextView textView;
        TextView mDateView;
        ImageView imageView;
        ConstraintLayout mMainView;
        QuickContactDivot mAvatarView;
        ImageButton imageButton;

        SentViewHolder(@NonNull View itemView) {
            super(itemView);
            mMainView = itemView.findViewById(R.id.main);
            mDateView = itemView.findViewById(R.id.date_view);
            textView = itemView.findViewById(R.id.text_view);
            imageView = itemView.findViewById(R.id.image_view);
            imageButton = itemView.findViewById(R.id.play_slideshow_button);
            mAvatarView = itemView.findViewById(R.id.avatar);
        }

        private void setupViewHolder(Data message) {
            mDateView.setText(message.getUpdated());
            mAvatarView.setVisibility(View.VISIBLE);
            daoRepository.getImageOfCharacter(message.getCharacter_id(), (uri, $completion) -> {
                Picasso.get().load(uri).into(mAvatarView);
                Drawable d = null;
                if (mAvatarView != null) {
                    d = mAvatarView.getDrawable();
                }
                if (d != null) {
                    mAvatarView.setImageDrawable(d);
                    mAvatarView.setVisibility(View.VISIBLE);
                } else {
                    mAvatarView.setImageDrawable(sDefaultContactImage);
                    mAvatarView.setVisibility(View.VISIBLE);
                }

                return null;
            });


            if (message.getContent() == null) {
                textView.setVisibility(View.GONE);
            } else {
                textView.setVisibility(View.VISIBLE);
            }

            imageView.setVisibility(View.GONE);
            imageButton.setVisibility(View.GONE);
            if (message.getContent() != null) {
                textView.setText(message.getContent());
            } else {
                textView.setText("");
            }
        }
    }

    class ReceiveViewHolder extends RecyclerView.ViewHolder {

        TextView textView;
        TextView mDateView;
        ImageView imageView;
        ConstraintLayout mMainView;
        QuickContactDivot mAvatarView;
        ImageButton imageButton;

        ReceiveViewHolder(@NonNull View itemView) {
            super(itemView);
            mMainView = itemView.findViewById(R.id.main);
            mDateView = itemView.findViewById(R.id.date_view);
            textView = itemView.findViewById(R.id.text_view);
            imageView = itemView.findViewById(R.id.image_view);
            imageButton = itemView.findViewById(R.id.play_slideshow_button);
            mAvatarView = itemView.findViewById(R.id.avatar);
        }

        private void setupViewHolder(Data message) {
            mDateView.setText(message.getUpdated());
            daoRepository.getImageOfCharacter(message.getCharacter_id(), (uri, $completion) -> {
                Picasso.get().load(uri).into(mAvatarView);
                Drawable d = null;
                if (mAvatarView != null) {
                    d = mAvatarView.getDrawable();
                }
                if (d != null) {
                    mAvatarView.setImageDrawable(d);
                    mAvatarView.setVisibility(View.VISIBLE);
                } else {
                    mAvatarView.setImageDrawable(sDefaultContactImage);
                    mAvatarView.setVisibility(View.VISIBLE);
                }

                return null;
            });


            textView.setVisibility(View.VISIBLE);
            imageView.setVisibility(View.GONE);
            imageButton.setVisibility(View.GONE);
            if (message.getContent() != null) {
                textView.setText(message.getContent());
            } else {
                textView.setText("");
            }


        }
    }
}

File: app/src/main/java/com/testosterolapp/unrd/messages/MessagesScreenAdapter.kt
package com.testosterolapp.unrd.messages

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import com.testosterolapp.unrd.R
import com.testosterolapp.unrd.data.Data
import com.testosterolapp.unrd.db.DaoRepository
import com.testosterolapp.unrd.db.DaoRepository.OnGetImageOfCharacter
import com.testosterolapp.unrd.util.QuickContactDivot
import java.util.*

open class MessagesScreenAdapter(mContext: Context, private val messagesList: ArrayList<Data>?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val daoRepository: DaoRepository = DaoRepository(mContext)

    override fun getItemViewType(position: Int): Int {
        if (messagesList != null) {
            if (position <= messagesList.size) {
                val message = messagesList[position]
                return if (message.character_id != null && message.character_id != MAIN_CHAR_ID) {
                    2
                } else {
                    TYPE_SENT
                }
            }
        }
        return TYPE_SENT
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): RecyclerView.ViewHolder {
        return if (i != TYPE_SENT) {
            //SENT UI
            val itemView = LayoutInflater.from(viewGroup.context)
                    .inflate(R.layout.message_list_sent_item, viewGroup, false)
            SentViewHolder(itemView)
        } else {
            //RECEIVE UI
            val itemView = LayoutInflater.from(viewGroup.context)
                    .inflate(R.layout.message_list_receive_item, viewGroup, false)
            ReceiveViewHolder(itemView)
        }
    }

    private fun getItem(position: Int): Data {
        return messagesList!![position]
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position <= -1) {
            return
        }
        val data: Data = try {
            getItem(position)
        } catch (e: IndexOutOfBoundsException) {
            return
        }
        if (holder is SentViewHolder) {
            holder.setupViewHolder(data)
        } else if (holder is ReceiveViewHolder) {
            holder.setupViewHolder(data)
        }
    }

    override fun getItemCount(): Int {
        return messagesList!!.size
    }

    internal inner class SentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var textView: TextView = itemView.findViewById(R.id.text_view)
        private var mDateView: TextView = itemView.findViewById(R.id.date_view)
        private var imageView: ImageView = itemView.findViewById(R.id.image_view)
        private var mAvatarView: QuickContactDivot? = itemView.findViewById(R.id.avatar)
        private var imageButton: ImageButton = itemView.findViewById(R.id.play_slideshow_button)
        fun setupViewHolder(message: Data) {
            mDateView.text = message.updated
            mAvatarView!!.visibility = View.VISIBLE
            daoRepository.getImageOfCharacter(message.character_id, object : OnGetImageOfCharacter {
                override suspend fun onComplete(uri: String?) {
                    Picasso.get().load(uri).into(mAvatarView)
                    var d: Drawable? = null
                    if (mAvatarView != null) {
                        d = mAvatarView!!.drawable
                    }
                    if (d != null) {
                        mAvatarView!!.setImageDrawable(d)
                        mAvatarView!!.visibility = View.VISIBLE
                    } else {
                        mAvatarView!!.setImageDrawable(sDefaultContactImage)
                        mAvatarView!!.visibility = View.VISIBLE
                    }
                }
            })
            if (message.content == null) {
                textView.visibility = View.GONE
            } else {
                textView.visibility = View.VISIBLE
            }
            imageView.visibility = View.GONE
            imageButton.visibility = View.GONE
            if (message.content != null) {
                textView.text = message.content
            } else {
                textView.text = ""
            }
        }

    }

    internal inner class ReceiveViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var textView: TextView = itemView.findViewById(R.id.text_view)
        private var mDateView: TextView = itemView.findViewById(R.id.date_view)
        private var imageView: ImageView = itemView.findViewById(R.id.image_view)
        private var mAvatarView: QuickContactDivot? = itemView.findViewById(R.id.avatar)
        private var imageButton: ImageButton = itemView.findViewById(R.id.play_slideshow_button)
        fun setupViewHolder(message: Data) {
            mDateView.text = message.updated
            daoRepository.getImageOfCharacter(message.character_id, object : OnGetImageOfCharacter {
                override suspend fun onComplete(uri: String?) {
                    Picasso.get().load(uri).into(mAvatarView)
                    var d: Drawable? = null
                    if (mAvatarView != null) {
                        d = mAvatarView!!.drawable
                    }
                    if (d != null) {
                        mAvatarView!!.setImageDrawable(d)
                        mAvatarView!!.visibility = View.VISIBLE
                    } else {
                        mAvatarView!!.setImageDrawable(sDefaultContactImage)
                        mAvatarView!!.visibility = View.VISIBLE
                    }
                }
            })
            textView.visibility = View.VISIBLE
            imageView.visibility = View.GONE
            imageButton.visibility = View.GONE
            if (message.content != null) {
                textView.text = message.content
            } else {
                textView.text = ""
            }
        }

    }

    companion object {
        private const val TYPE_SENT = 1
        private const val MAIN_CHAR_ID: Long = 218
        private var sDefaultContactImage: Drawable? = null
    }

    init {
        if (sDefaultContactImage == null) {
            sDefaultContactImage = mContext.resources.getDrawable(R.drawable.ic_contact_picture)
        }
    }
}